clouddriver-appengine/src/main/groovy/com/netflix/spinnaker/clouddriver/appengine/deploy/ops/DeployAppEngineAtomicOperation.groovy

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2016 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.appengine.deploy.ops
+
+import com.netflix.spinnaker.clouddriver.appengine.deploy.AppEngineMutexRepository
+import com.netflix.spinnaker.clouddriver.appengine.deploy.AppEngineServerGroupNameResolver
+import com.netflix.spinnaker.clouddriver.appengine.deploy.description.DeployAppEngineDescription
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.deploy.DeploymentResult
+import com.netflix.spinnaker.clouddriver.jobs.JobExecutor
+import com.netflix.spinnaker.clouddriver.jobs.JobRequest
+import com.netflix.spinnaker.clouddriver.jobs.JobStatus
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import org.springframework.beans.factory.annotation.Autowired
+
+class DeployAppEngineAtomicOperation implements AtomicOperation<DeploymentResult> {
+  private static final String BASE_PHASE = "DEPLOY"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  @Autowired
+  JobExecutor jobExecutor
+
+  DeployAppEngineDescription description
+
+  DeployAppEngineAtomicOperation(DeployAppEngineDescription description) {
+    this.description = description
+  }
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack", "freeFormDetails": "details", "repositoryUrl": "https://github.com/organization/project.git", "branch": "feature-branch", "credentials": "my-appengine-account", "appYamlPath": "path/to/app.yaml" } } ]' "http://localhost:7002/appengine/ops"
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack", "freeFormDetails": "details", "repositoryUrl": "https://github.com/organization/project.git", "branch": "feature-branch", "credentials": "my-appengine-account", "appYamlPath": "path/to/app.yaml", "promote": true, "stopPreviousVersion": true } } ]' "http://localhost:7002/appengine/ops"
+   */
+  @Override
+  DeploymentResult operate(List priorOutputs) {
+    def directoryName = getLegalDirectoryName(description.repositoryUrl)
+
+    /*
+    * We can't allow concurrent deploy operations on the same local repository.
```

Comment:
We talked about this offline. The locking is OK, since each clouddriver instance would clone the repository independently.